### ISSUE TITLE
l10n export is broken (Fatal error: Unexpected token (93:10))

### ISSUE THREAD

**Issue by flodolo** [mozilla/fxa#13893]
`npx grunt l10n-extract` is currently broken, so there's no way to extract strings for localization.

This is the relevant section of the log ([full log here](https://github.com/mozilla/fxa-content-server-l10n/runs/7681667280?check_suite_focus=true)). Tried re-running the job to check if it was a temporary issue, but it's not.

CC @clouserw @bcolsson 

```
Running "jsxextract" task
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
CRITICAL fxa-content-server.uncaughtException: SyntaxError: Unexpected token (93:10)
    at Parser.pp$4.raise (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:2757:13)
    at Parser.pp.unexpected (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:647:8)
    at Parser.pp$3.parseExprAtom (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:2196:10)
    at Parser.parseExprAtom (node_modules/acorn-jsx/inject.js:399:24)
    at Parser.pp$3.parseExprSubscripts (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:2047:19)
    at Parser.pp$3.parseMaybeUnary (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:2024:17)
    at Parser.pp$3.parseExprOps (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1966:19)
    at Parser.pp$3.parseMaybeConditional (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1949:19)
    at Parser.pp$3.parseMaybeAssign (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1925:19)
    at Parser.pp$3.parseMaybeConditional (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1954:28)
    at Parser.pp$3.parseMaybeAssign (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1925:19)
    at Parser.pp$1.parseVar (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1170:26)
    at Parser.pp$1.parseVarStatement (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1034:8)
    at Parser.pp$1.parseStatement (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:788:17)
    at Parser.pp$1.parseBlock (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1112:23)
    at Parser.pp$1.parseTryStatement (node_modules/acorn-jsx/node_modules/acorn/dist/acorn.js:1007:21)
Fatal error: Unexpected token (93:10)


Execution Time (2022-08-04 22:12:40 UTC-0)
loading tasks   1.3s  ▇▇▇▇▇ 10%
jsxextract     11.5s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 90%
Total 12.9s

Error: Process completed with exit code 3.
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXA-5679)

**Comment by flodolo** [comment]
It can be reproduced locally in `packages/fxa-content-server`, running `npx grunt l10n-extract`.

So far I've been able to narrow it down to a JS file in [this function](https://github.com/mozilla/fxa/blob/main/packages/fxa-content-server/grunttasks/l10n-extract.js#L64-L84), which points to [this commit](https://github.com/mozilla/fxa/commit/c7973815e3e4e5e2ce4f39f6caec30e23e1575c7) (cc @dschom). Still not sure what's wrong though, or if the analysis is correct.

**Comment by flodolo** [comment]
> which points to [this commit](https://github.com/mozilla/fxa/commit/c7973815e3e4e5e2ce4f39f6caec30e23e1575c7) (cc @dschom). Still not sure what's wrong though, or if the analysis is correct.

This doesn't seem correct. Or, at least, now I'm trying again to checkout previous commits, and they fail (they were working before). No clue what's happening at this point :-(

**Comment by flodolo** [comment]
It's failing on [these question marks](https://github.com/mozilla/fxa/blob/c7973815e3e4e5e2ce4f39f6caec30e23e1575c7/packages/fxa-content-server/server/lib/sentry.js#L93-L100). Unsupported syntax by the parser? If I'm looking at the right package, that's very old and pretty much unmaintained.

**Comment by data-sync-user** [comment]
➤ Wil Clouser commented:

https://github.com/mozilla/fxa/blob/main/packages/fxa-content-server/grunttasks/l10n-extract.js ( https://github.com/mozilla/fxa/blob/main/packages/fxa-content-server/grunttasks/l10n-extract.js|smart-link )  is the extraction script which hasn’t changed in two years.  It looks like it’s using jsxgettext-recursive-next which also hasn’t been upgraded in years (and was also written by ex-mozilla people).

I’m not sure when optional chaining support was added, but maybe that library can’t handle it.  Francesco Lodolo is there a library that is standard at mozilla for this or is the answer “stop using gettext” ? 

/cc Dan Schomburg

**Comment by data-sync-user** [comment]
➤ Francesco Lodolo commented:

Unfortunately I’m not aware of any library. Might be worth checking with AMO folks to ask what they’re using? That’s the largest project left still using GetText at this point.

P.S. I’m very confused by that package. npm ( https://www.npmjs.com/package/jsxgettext-recursive-next ) has a version (1.1.0), which is the one used in FxA, that doesn’t actually exist in the repo ( https://github.com/zaach/jsxgettext-recursive ) (which has 1.0.1)?

**Comment by data-sync-user** [comment]
➤ Wil Clouser commented:

Looks like they are using Django: https://github.com/mozilla/addons-server/blob/master/scripts/extract-l10n.sh ( https://github.com/mozilla/addons-server/blob/master/scripts/extract-l10n.sh|smart-link )

**Comment by pdehaan** [comment]
Per https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining, it looks like optional chaining syntax was added in Node 14.0+. But from the looks of your log output, you're already using 14.2:
> _"Found in cache @ /opt/hostedtoolcache/node/14.20.0/x64"_

Although I think the current recommended/minimum Node version for FxA is 16, per: https://github.com/mozilla/fxa/blob/2573c0d64c932ee4be96d4200710336557008819/package.json#L44 

But I too am rather curious why this would suddenly break/change _now_, when it looks like a lot of the code/process is pretty old.

**Comment by data-sync-user** [comment]
➤ Francesco Lodolo commented:

The Grunt task ( https://github.com/mozilla/fxa/blob/main/packages/fxa-content-server/grunttasks/l10n-extract.js#L52-L78 ) and parser code might be ancient, but the code that introduced the optional chaining is 6 days old ( https://github.com/mozilla/fxa/commit/c7973815e3e4e5e2ce4f39f6caec30e23e1575c7 ) (per previous comments).

The explanation seems very simple to me: this syntax was never used yet in a .js file within the packages/fxa-content-server folder (the rest of the repo doesn't matter).

**Comment by data-sync-user** [comment]
➤ Francesco Lodolo commented:

Possible work-around, given we know the long term solution is to move everything to Fluent:

  * Rewrite that code to avoid optional chaining.
  * Add a workflow that actually runs the extraction when affected files are touched in the fxa-content-server path.

The latter is to know when the workflow fails, without waiting for the fail to happen in the l10n repo.

**Comment by data-sync-user** [comment]
➤ Wil Clouser commented:

There is already a patch for this under review (see the “development” section on the right in jira)

**Comment by data-sync-user** [comment]
➤ Bryan Olsson commented:

Noting here I noticed the PR was merged and tested our string extract workflow, which ran successfully.

**Comment by pdehaan** [comment]
> gettext relies on a JavaScript parser called acorn ( [acornjs/acorn](https://github.com/acornjs/acorn) ) which doesn’t support optional chaining and resulted in the error we see here.

Ah, OK, I think I understand now. Looks like acorn is currently v8.8.0 and chaining operator _has_ been supported since acorn@7.3.0 (per https://github.com/acornjs/acorn/issues/1010/) but jsxgettext@0.11.0 is pinning to ^5.1.2 which resolves to acorn@5.7.4; which checks out versus https://semver.npmjs.com/:

```sh
yarn why acorn | grep -A1 jsxgettext
├─ jsxgettext@npm:0.11.0
│  └─ acorn@npm:5.7.4 (via npm:^5.1.2)

npm info acorn time | grep 5.7.4
#  '5.7.4': '2020-03-09T23:25:56.153Z',
```
